### PR TITLE
Remove `PartIDTag` from nodes in partition 

### PR DESCRIPTION
**Pull request by inducer**
See #307.

- [x] May need a pytools release for https://github.com/inducer/pytools/pull/126: https://pypi.org/project/pytools/2022.1.4/

Best read commit-by-commit.

### CODE REVIEW

**Comment by inducer** [comment]
@kaushikcfd This is now passing and ready for a look.

File: pytato/distributed.py [hunk 1]
```diff
@@ -793,6 +829,15 @@ def rec(self,  # type: ignore[override]
             return result
 
 
+def _remove_part_id_tag(ary: ArrayOrNames) -> Array:
+    assert isinstance(ary, Array)
+
+    # FIXME This spurious assignment helps mypy understand this code.
+    # I wasn't able to reproduce this in isolation.
```
Comment (kaushikcfd) [inline pytato/distributed.py:836]: I don't think this particular comment line is too helpful but has the potential to confuse the reader.

**Review by kaushikcfd — APPROVED** [review]
LGTM, thanks! I had completely missed nested distributed comm nodes being a thing.

File: pytato/distributed.py (hunk 1 above)
Comment (inducer) [reply pytato/distributed.py:836]: Could you explain? What I'm trying to do is explain why this doesn't say `return stuff`, the reason being a (perceived) mypy bug. Would you prefer I just get rid of the comment?

**Comment by inducer** [comment]
>  I had completely missed nested distributed comm nodes being a thing.

To be clear, this is about executing one (distributed) workload that uses results from another, and then being confused by leftover partitioning tags. Maybe that's what you meant by "nesting", but I figured I'd clarify just to be sure.

**Comment by kaushikcfd** [comment]
> Maybe that's what you meant by "nesting", but I figured I'd clarify just to be sure.

Yep, that's what I meant.

File: pytato/distributed.py (hunk 1 above)
Comment (inducer) [reply pytato/distributed.py]: Turns out I was just being dumb about reproducing it. It reproduces just fine: https://github.com/python/mypy/issues/12626

Comment (inducer) [reply pytato/distributed.py]: (now linked)

**Comment by inducer** [comment]
Thanks for taking a look!

**Pull request merged**